fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -458,8 +475,417 @@ Value makePadding(int size) {
+		self->wPtr += BlobCipherEncryptHeader::headerSize;
+		self->dataPayloadStart = self->wPtr;
+
+		// If this is NOT the first block then write duplicate stuff needed from last block
+		if (self->blockEnd > self->blockSize) {
+			appendStringRefWithLenToBuffer(self, &lastKey);
+			appendStringRefWithLenToBuffer(self, &self->lastKey);
+			if (writeValue) {
+				appendStringRefWithLenToBuffer(self, &self->lastValue);
+			}
+		}
+
+		// There must now be room in the current block for bytesNeeded or the block size is too small
+		if (expectedFileSize(self) + bytesNeeded > self->blockEnd) {
+			throw backup_bad_block_size();
+		}
+
+		return Void();
+	}
+
+	Future<Void> padEnd(bool final) {
+		if (expectedFileSize(this) > 0) {
+			return newBlock(this, 0, StringRef(), true, final);
+		}
+		return Void();
+	}
+
+	// Ends the current block if necessary based on bytesNeeded.
+	ACTOR static Future<Void> newBlockIfNeeded(EncryptedRangeFileWriter* self, int bytesNeeded) {
+		if (expectedFileSize(self) + bytesNeeded > self->blockEnd) {
+			wait(newBlock(self, bytesNeeded, self->lastKey, true));
+		}
+		return Void();
+	}
+
+	ACTOR static Future<bool> finishCurTenantBlockStartNewIfNeeded(EncryptedRangeFileWriter* self,
+	                                                               Key k,
+	                                                               Value v,
+	                                                               bool writeValue) {
+		// Don't want to start a new block if the current key or previous key is empty
+		if (self->lastKey.size() == 0 || k.size() == 0) {
+			return false;
+		}
+		state std::pair<int64_t, TenantName> curKeyTenantInfo = wait(getTenantIdName(k, self->tenantCache));
+		state std::pair<int64_t, TenantName> prevKeyTenantInfo =
+		    wait(getTenantIdName(self->lastKey, self->tenantCache));
+		// crossing tenant boundaries so finish the current block using only the tenant prefix of the new key
+		if (curKeyTenantInfo.first != prevKeyTenantInfo.first) {
+			CODE_PROBE(true, "crossed tenant boundaries");
+			state KeyRef endKey = k;
+			// If we are crossing a boundary with a key that has a tenant prefix then truncate it
+			if (curKeyTenantInfo.first != SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID &&
+			    curKeyTenantInfo.first != FDB_DEFAULT_ENCRYPT_DOMAIN_ID) {
+				endKey = StringRef(k.begin(), 8);
+			}
+			state ValueRef newValue = StringRef();
+			self->lastKey = k;
+			self->lastValue = v;
+			appendStringRefWithLenToBuffer(self, &endKey);
+			appendStringRefWithLenToBuffer(self, &newValue);
+			wait(newBlock(self, 0, endKey, writeValue));
+			wait(updateEncryptionKeys(self, self->lastKey, self->tenantCache));
+			return true;
+		}
+		return false;
+	}
+
+	// Start a new block if needed, then write the key and value
+	ACTOR static Future<Void> writeKV_impl(EncryptedRangeFileWriter* self, Key k, Value v) {
+		if (!self->cipherKeys.headerCipherKey.isValid() || !self->cipherKeys.textCipherKey.isValid()) {
```

Comment:
Is this `if` condition possible? I would have assume `assert textCipherKey.isValid () && headerCipherKey.isValid() && iv.size() == AES_256_IV_LENGTH` ?

Reply:
Sometimes the very first key in a file is empty string, so then we will not fetch encryption keys in WriteKey_impl so we will need to fetch it here